modem: hl78xx: Add AT shell and AirVantage FOTA support

zafersn opened the pull request:
This PR enhances the HL78xx modem driver with two key features:

1. AT Shell Support – Enables sending AT commands directly from the Zephyr shell for testing, debugging, and modem interaction.
2. Sierra Wireless AirVantage FOTA Support – Adds integration for firmware-over-the-air (FOTA) updates via the AirVantage platform.

- Introduced AT shell interface to allow developers to issue AT commands through Zephyr’s shell subsystem.
- Added AirVantage FOTA support to enable remote firmware management for HL78xx-based devices.
- Updated driver logic and configuration options to ensure compatibility with recent firmware versions.
- Verified functionality on the Pinnacle DK board (tested FOTA via AirVantage and AT shell features).

These additions improve both development workflow (through shell access) and device maintainability (through AirVantage-based updates).

rerickson1 commented on drivers/modem/hl78xx/hl78xx_cfg.c:
```diff
@@ -192,13 +214,41 @@ int hl78xx_set_apn_internal(struct hl78xx_data *data, const char *apn, uint16_t
+		goto error;
+	}
+#endif /* CONFIG_MODEM_HL78XX_AIRVANTAGE */
 	data->status.apn.state = APN_STATE_CONFIGURED;
```
Why do this when setting the APN?

zafersn replied:
That’s because the PDP context is being configured there.
From my understanding, based on the HL78xx FOTA Guidelines r5 (Document 2174223):

> 4.1.1 PDP Context
> The module uses UDP (CoAP) to communicate with AirVantage to create DM sessions.
> A PDP context can be assigned for this connection using AT+WDSS=2,<cid>, where <cid> is the PDP context
> identifier as configured with AT+CGDCONT. The change initiated by AT+WDSS=2,<cid> only takes effect while
> syncing with the AVMS server. It does not affect the FOTA package download. For HL780x, HL780x always uses
> CID1 for FOTA whatever the CID setting in +WDSS command is.
> Send AT+WDSS? to check the current APN settings. If an APN has not been specified, CME Error 650 will occur when
> trying to initiate a connection.

rerickson1 commented on drivers/modem/hl78xx/Kconfig.hl78xx:
```diff
@@ -57,33 +57,21 @@ config MODEM_HL78XX_12_FW_R6
+config MODEM_AT_SHELL_RESPONSE_MAX_SIZE
+	int "Maximum size of AT response"
+	default 64
+endif # MODEM_HL78XX_AT_SHELL
```
Fix some formatting here. Should be blank lines between separate configs and the start of the if.


```suggestion
config MODEM_HL78XX_AT_SHELL
	bool "AT command shell"
	help
	  Enable AT command shell for the modem.

if MODEM_HL78XX_AT_SHELL

config MODEM_AT_SHELL_RESPONSE_TIMEOUT_MS
	int "Timeout waiting for response to AT command in milliseconds"
	default 500

config MODEM_AT_SHELL_COMMAND_MAX_SIZE
	int "Maximum size of AT command"
	default 32

config MODEM_AT_SHELL_RESPONSE_MAX_SIZE
	int "Maximum size of AT response"
	default 64

endif # MODEM_HL78XX_AT_SHELL
```

rerickson1 reviewed (dismissed):
AT shell works great on Pinnacle 100 DVK.
Did not test AirVantage FOTA, but I trust @zafersn did :)

zafersn commented:
> AT shell works great on Pinnacle 100 DVK. Did not test AirVantage FOTA, but I trust @zafersn did :)

Always! 🙂

Here’s a snapshot from a FOTA over AirVantage that I ran on the dev kit:

<img width="808" height="427" alt="image" src="https://github.com/user-attachments/assets/690247e3-fc57-4380-88d6-28ef80ce6ca0" />

The thing is, we don’t have many firmware version options available on AirVantage for upgrading the modem unless you specifically request them from Sierra Wireless. So, HL7800.4.7.1.1 (18.09.6.ALT1250.rc12) was the only firmware version available to test on the HL7800. As far as I remember, I had to downgrade the existing modem firmware to make it compatible with the Delta FOTA process.

JarmouniA commented on drivers/modem/hl78xx/hl78xx.c:
```diff
@@ -1227,7 +1382,118 @@ static int hl78xx_on_carrier_on_state_leave(struct hl78xx_data *data)
+	hl78xx_stop_timer(data);
+	return 0;
+}
+#endif /* CONFIG_MODEM_HL78XX_AIRVANTAGE */
```
Add an empty line between functions.

zafersn replied:
done

JarmouniA reviewed (dismissed):
Please put sample changes in a separate commit.

zafersn commented on drivers/modem/hl78xx/hl78xx.c:
```diff
@@ -967,11 +1097,10 @@ static void hl78xx_run_init_fail_script_event_handler(struct hl78xx_data *data,
+				"mode is disabled. If you’re unsure, enable it by adding the "
+				"corresponding DTS configuration entry.");
 		}
-
```
done

zafersn commented on drivers/modem/hl78xx/hl78xx.h:
```diff
@@ -68,38 +69,43 @@
-#define DEACTIVATE_PDP_CONTEXT             "AT+CGACT=0"
-#define ACTIVATE_PDP_CONTEXT               "AT+CGACT=1"
+#define DEACTIVATE_PDP_CONTEXT                   "AT+CGACT=0"
+#define ACTIVATE_PDP_CONTEXT                     "AT+CGACT=1"
```
please refer to my comment above for this issue

zafersn commented on drivers/modem/hl78xx/hl78xx.h:
```diff
@@ -254,7 +279,34 @@ struct hl78xx_network_operator {
 	char operator[MDM_MODEL_LENGTH];
 	uint8_t format;
 };
-
```
done

zafersn commented on drivers/modem/hl78xx/hl78xx_chat.h:
```diff
@@ -47,13 +48,25 @@ int hl78xx_run_init_fail_script_async(struct hl78xx_data *data);
 
 /* Getter for ksrat match (moved into chat TU) */
 const struct modem_chat_match *hl78xx_get_ksrat_match(void);
-
```
done

JordanYates reviewed (changes requested):
Not a user of the AT shell or HL78xx, so I can't speak to the content of the changes.
Requests are for docs and commit clarity

rerickson1 commented on drivers/modem/hl78xx/Kconfig.hl78xx:
```diff
@@ -719,6 +717,59 @@ config MODEM_MIN_ALLOWED_SIGNAL_STRENGTH
 	    default -140
 	    range - 140 0
 
+config MODEM_HL78XX_AIRVANTAGE
```
Add this to the 4.4 release notes now. 4.3 is soon to be released. This PR will not go in 4.3

JordanYates reviewed (changes requested):
The renamed Kconfig options are missing from the migration guide.

zafersn commented on drivers/modem/hl78xx/Kconfig.hl78xx:
```diff
@@ -759,45 +742,73 @@ config MODEM_HL78XX_NUM_SOCKETS
+config MODEM_HL78XX_DEV_STARTUP_TIME_MS
 	int "Wait before assuming the device is ready."
-	default 1000
+	default 120
```
@JordanYates FYI, doc/releases/migration-guide-4.4.rst and doc/releases/release-notes-4.4.rst have been updated

zafersn replied:
Ok. I have updated as you recommended. How does it look now?

zafersn replied:
@JordanYates  will you be able to review again?

zafersn committed:
drivers: modem: hl78xx: reduce startup_ms and increase init chat timeout

The previous delay could cause missed URCs because the pipe and chat
layer were not yet attached when the modem started sending its first
messages (+KSUP). The modem boot takes ~1.8s, while the driver only
needs ~120ms to set up the interfaces, so reducing the delay ensures
the receiver is ready before the modem outputs any URCs.

Increase hl78xx_init_chat_script command limit from 10 to 100 to allow
longer initialization sequences.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
sample: hl78xx: Add AirVantage FOTA support

Demonstrate SWIR AirVantage FOTA in the HL78xx sample.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

JarmouniA reviewed (commented):
There are still many formatting changes mixed with code changes/additions in the same commit, please fix.
https://github.com/zephyrproject-rtos/zephyr/pull/97919#discussion_r2510457141
https://github.com/zephyrproject-rtos/zephyr/pull/97919#discussion_r2538808383

JarmouniA commented on drivers/modem/hl78xx/hl78xx.h:
```diff
@@ -68,38 +69,39 @@
-#define DEACTIVATE_PDP_CONTEXT             "AT+CGACT=0"
-#define ACTIVATE_PDP_CONTEXT               "AT+CGACT=1"
+#define DEACTIVATE_PDP_CONTEXT                   "AT+CGACT=0"
+#define ACTIVATE_PDP_CONTEXT                     "AT+CGACT=1"
```
formatting change

JarmouniA commented on include/zephyr/drivers/modem/hl78xx_apis.h:
```diff
@@ -18,26 +18,26 @@ extern "C" {
+#define MDM_ICCID_LENGTH         22
+#define MDM_APN_MAX_LENGTH       64
+#define MDM_MAX_CERT_LENGTH      4096
+#define MDM_MAX_HOSTNAME_LEN     128
```
formatting change

JarmouniA replied:
You probably didn't push the latest version, the 1st commit still has everything.

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
Release Notes and Migration Guide additions should be in a separate commit.

JarmouniA commented:
> were applied automatically by clang-format.

Already responded to this
https://github.com/zephyrproject-rtos/zephyr/pull/97919#discussion_r2510457141
https://github.com/zephyrproject-rtos/zephyr/pull/97919#discussion_r2538808383

> So there isn’t much I can control in that regard.

You have full control over what goes in a commit, and choosing to run Clang-format automatically & to keep all the changes is your problem, not the reviewers'.

zafersn committed:
drivers: modem: hl78xx: Add AT Shell support

Adds AT shell support, enabling developers to send AT commands directly.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
doc: releases: Document HL78xx modem updates for 4.4

Update migration guide for HL78xx Kconfig option renames and startup
time changes. Add release notes for new AT Shell and AirVantage FOTA
support.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
drivers: modem: hl78xx: reorder Kconfig entries, and comment fixes

This patch applies multiple small cleanups across the HL78xx modem
driver, improving readability, consistency, and diagnostics without
changing logic.

Key changes:
- Added missing blank lines in Kconfig sections for readability.
- Improved several log messages for clarity and guidance to users.
- Added a debug log when RAT/Band configuration requires a modem restart.
- Replaced an incorrect SET_RAT_GMS_CMD macro name with SET_RAT_GSM_CMD.
- Added TODO marker for unhandled script failure case.
- Corrected log format string in event handler to include event ID.
- Added an informational log on modem initialization.
- Added clarifying comments in header structures and APN lookup path.
- Normalized #endif comment style throughout the driver.
- Fixed use of CME_ERROR_STRING in chat match definitions.
- Applied formatting hints (clang-format off/on) around multi-line
  MODEM_CHAT_MATCHES blocks.
- Improved enum documentation in public API header.

These changes improve maintainability and developer usability without
modifying driver behavior.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
drivers: modem: hl78xx: Fix GSM RAT registration and PDP handling

Fix several issues in GSM mode for the HL78xx modem driver to
ensure correct registration reporting, RAT handling, and PDP
context activation.

Changes include:
- Skip band configuration when RAT=GSM to prevent +CME ERROR
  responses.
- Explicit PDP activation (AT+CGACT=1,1) for GSM data sessions.
- Parse +CGACT URCs and track PDP activation state internally.
- Introduce gprs status struct and update initialization
  scripts to handle GSM reliably.
- Add helper functions and chat scripts to manage GSM
  registration and PDP flow.
- Correct GSM registration-status handling: enable CREG=3 and
  disable unused CEREG.
- Dynamically switch between CREG (GSM) and CEREG (LTE/NB-IoT)
  based on active RAT.

These changes improve driver stability and ensure reliable data
and registration behavior on GSM networks.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
drivers: modem: hl78xx: Improve restart handling and boot tracking

This update refactors the HL78xx initialization and runtime state
machine to provide robust handling of unexpected modem restarts
and improve PDP and boot status management.

Key changes:
- Add KSUP unsolicited response to the global chat match table
  so modem boot status is monitored independently from the
  initialization script.
- Introduce a boot-status structure tracking previous boot state
  to detect unexpected restarts.
- Dispatch a new MODEM_HL78XX_EVENT_MDM_RESTART event when the
  modem reports a READY status after having been previously
  booted.
- Add restart-handling transitions across relevant states
  (await-power-on, RAT-config, await-registered, carrier-on) to
  ensure the driver cleanly re-enters the initialization
  sequence.
- Rework init and post-restart script entry points to run
  asynchronously via hl78xx_run_init_script_async() and
  hl78xx_run_post_restart_script_async().
- Move bus-open and chat-attach operations out of the
  init-script state and into the await-power-on state for
  clearer separation of responsibilities.
- Expose a getter for the AT-ready match structure.

These changes improve resilience during modem restarts, clarify
the state machine flow, and prepare the driver for multi-PDP-
context configurations.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
drivers: modem: hl78xx: rename timing configs to clarify units

Rename HL78XX device timing configuration options to include units (ms):
- CONFIG_MODEM_HL78XX_DEV_POWER_PULSE_DURATION ->
  CONFIG_MODEM_HL78XX_DEV_POWER_PULSE_DURATION_MS
- CONFIG_MODEM_HL78XX_DEV_RESET_PULSE_DURATION ->
  CONFIG_MODEM_HL78XX_DEV_RESET_PULSE_DURATION_MS
- CONFIG_MODEM_HL78XX_DEV_STARTUP_TIME ->
  CONFIG_MODEM_HL78XX_DEV_STARTUP_TIME_MS
- CONFIG_MODEM_HL78XX_DEV_SHUTDOWN_TIME ->
  CONFIG_MODEM_HL78XX_DEV_SHUTDOWN_TIME_MS

Update MODEM_HL78XX_DEFINE_INSTANCE macro to use *_MS options.

This improves configuration clarity and robustness of modem init.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>

zafersn committed:
drivers: modem: hl78xx: Add AirVantage FOTA support

Sierra Wireless AirVantage FOTA updates in the HL78xx modem driver.

Signed-off-by: Zafer SEN <zafersn93@gmail.com>
